cumulusci/utils/fileutils.py

```diff
@@ -86,6 +92,206 @@ def load_from_source(
             yield path, f
 
 
+def cleanup_org_cache_dirs(keychain, project_config):
+    """Cleanup directories that are not associated with a connected/live org."""
+    domains = set()
+    for org in keychain.list_orgs():
+        org_config = keychain.get_org(org)
+        domain = org_config.get_domain()
+        if domain:
+            domains.add(domain)
+
+    assert project_config.project_cache_dir
+    assert keychain.global_config_dir
+
+    project_org_directories = (project_config.project_cache_dir / "orginfo").glob("*")
+    global_org_directories = (keychain.global_config_dir / "orginfo").glob("*")
+
+    for directory in list(project_org_directories) + list(global_org_directories):
+        if directory.name not in domains:
+            rmtree(directory)
+
+
+def proxy(funcname):
+    def func(self, *args, **kwargs):
+        real_func = getattr(self.fs, funcname)
+        return real_func(self.filename, *args, **kwargs)
+
+    func.__doc__ = getattr(base.FS, funcname).__doc__
+    return func
+
+
+class FSResource:
+    """Generalization of pathlib.Path to support S3, FTP, etc
+
+    Create them through the open_fs_resource module function or static
+    function which will create a context manager that generates an FSResource.
+
+    If you don't need the resource management aspects of the context manager,
+    you can call the `new()` classmethod."""
```

Comment:
What's the reason for inventing a new way to construct it instead of using `__init__`?

Reply:
Because `__init__`'ing the object is easy but discouraged. I put a small amount of effort in front of it so you're more likely to read the docs before you do it. And raising an exception in `__init__` increases the likelihood of you reading the docs and using the class correctly.

Some libraries rename the class (e.g. lxml.etree._Element) but then isinstance checks and repr get ugly. I thought that this is a better approach.

Reply:
What I'm trying to avoid is the situation we're in with SQL Alchemy/SQLite where we should have managed the engine as a resource but it wasn't obvious that we should have done so until we had tons of code that doesn't do so.

Reply:
Okay, you convinced me.